qiskit/primitives/containers/observable.py

```diff
@@ -0,0 +1,164 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2024.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+
+"""
+Sparse container class for an Estimator observable.
+"""
+from __future__ import annotations
+
+from typing import Union, Mapping as MappingType
+from collections.abc import Mapping
+from collections import defaultdict
+from numbers import Complex
+
+from qiskit.quantum_info import Pauli, SparsePauliOp
+
+
+ObservableKey = "tuple[tuple[int, ...], str]"  # pylint: disable = invalid-name
+ObservableKeyLike = Union[Pauli, str, ObservableKey]
+ObservableLike = Union[
+    SparsePauliOp,
+    ObservableKeyLike,
+    MappingType[ObservableKeyLike, complex],
+]
+"""Types that can be natively used to construct a :const:`Observable`."""
+
+
+class Observable(Mapping):
+    """A sparse container for a observable for an :class:`.Estimator` primitive."""
+
+    __slots__ = ("_data", "_num_qubits")
+
+    def __init__(
+        self,
+        data: Mapping[ObservableKey, complex],
+        num_qubits: int = None,
+        validate: bool = True,
+    ):
+        """Initialize an observables array.
+
+        Args:
+            data: The observable data.
+            num_qubits: The number of qubits in the data.
+            validate: If ``True``, the input data is validated during initialization.
+
+        Raises:
+            ValueError: If ``validate=True`` and the input observable-like is not valid.
+        """
+        self._data = data
```

Comment:
If we care about not sharing a reference to the internal thing:
```suggestion
        self._data = dict(data.items())
```

Reply:
Its a trade off with that vs performance since this requires an iteration over the data at init